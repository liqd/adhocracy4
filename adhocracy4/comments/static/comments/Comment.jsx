var ReportModal = require('../../../reports/static/reports/react_reports').ReportModal
var RatingBox = require('../../../ratings/static/ratings/react_ratings').RatingBox
var Modal = require('../../../static/Modal')
var CommentEditForm = require('./CommentEditForm')
var CommentForm = require('./CommentForm')
var CommentManageDropdown = require('./CommentManageDropdown')

var React = require('react')
var PropTypes = require('prop-types')
var django = require('django')

var safeHtml = function (text) {
  return { __html: text }
}

var localeDate = function (dateStr) {
  return new Date(dateStr).toLocaleString(document.documentElement.lang)
}

var getViewRepliesText = function (number, hide) {
  var fmts
  if (hide) {
    fmts = django.ngettext('hide one reply', 'hide %s replies', number)
  } else {
    fmts = django.ngettext('view one reply', 'view %s replies', number)
  }
  return django.interpolate(fmts, [number])
}

const answerTag = django.gettext('Answer')
const answerPlaceholderText = django.gettext('Your reply here')
const reportText = django.gettext('You want to report this content? Your message will be sent to the moderation. The moderation will look at the reported content. The content will be deleted if it does not meet our discussion rules (netiquette).')

class Comment extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      edit: false,
      showChildComments: false,
      replyFormHasFocus: false
    }
  }

  toggleEdit (e) {
    if (e) {
      e.preventDefault()
    }
    var newEdit = !this.state.edit
    this.setState({ edit: newEdit })
  }

  toggleShowComments (e) {
    e.preventDefault()
    var newShowChildComment = !this.state.showChildComments
    this.setState({
      showChildComments: newShowChildComment,
      replyFormHasFocus: false
    })
  }

  replyComments (e) {
    e.preventDefault()
    this.setState({
      showChildComments: true,
      replyFormHasFocus: true
    })
  }

  allowForm () {
    return !this.props.isReadOnly && this.props.content_type !== this.context.comments_contenttype
  }

  isOwner () {
    return this.props.user_name === this.context.user_name
  }

  renderRatingBox () {
    if (!this.props.is_deleted) {
      return (
        <RatingBox
          contentType={this.context.comments_contenttype}
          objectId={this.props.id}
          authenticatedAs={this.context.isAuthenticated ? this.context.user_name : null}
          positiveRatings={this.props.positiveRatings}
          negativeRatings={this.props.negativeRatings}
          userRating={this.props.userRating}
          userRatingId={this.props.userRatingId}
          isReadOnly={this.props.isReadOnly}
        />
      )
    }
  }

  renderModeratorLabel () {
    const moderatorTag = django.gettext('Moderator')
    if (this.props.authorIsModerator && !this.props.is_deleted) {
      return (
        <span className="label label-subtle">{moderatorTag}</span>
      )
    }
  }

  renderLastDate () {
    const lastEditText = django.gettext('Latest edit on')
    let lastDate
    if (this.props.modified === null || this.props.is_deleted) {
      lastDate = localeDate(this.props.created)
    } else {
      lastDate = lastEditText + ' ' + localeDate(this.props.modified)
    }
    return (
      <span className="comment-submission-date">{lastDate}</span>
    )
  }

  renderComment () {
    let comment
    if (this.state.edit) {
      comment = (
        <CommentEditForm
          subjectType={this.props.content_type}
          subjectId={this.props.object_pk}
          comment={this.props.children}
          error={this.props.editError}
          errorMessage={this.props.errorMessage}
          handleErrorClick={() => this.props.handleEditErrorClick(this.props.index, this.props.parentIndex)}
          rows="5"
          handleCancel={this.toggleEdit.bind(this)}
          onCommentSubmit={newComment => {
            this.props.handleCommentModify(newComment, this.props.index, this.props.parentIndex)
              .then(() => {
                this.setState({
                  edit: false
                })
              })
          }}
        />
      )
    } else {
      comment = <div className="comment-text" dangerouslySetInnerHTML={safeHtml(this.props.children)} />
    }
    return comment
  }

  renderDeleteModal () {
    const confirmDeleteText = django.gettext('Do you really want to delete this comment?')
    const deleteTag = django.gettext('Delete')
    const abortTag = django.gettext('Abort')
    if (this.isOwner() || this.context.isModerator) {
      return (
        <Modal
          name={`comment_delete_${this.props.id}`}
          partials={{ title: confirmDeleteText }}
          submitHandler={() => this.props.handleCommentDelete(this.props.index, this.props.parentIndex)}
          action={deleteTag}
          abort={abortTag}
          btnStyle="cta"
        />
      )
    }
  }

  render () {
    const CommentList = require('./CommentList')

    return (
      <div className="comment">
        <ReportModal
          name={`report_comment_${this.props.id}`}
          title={reportText}
          btnStyle="cta"
          objectId={this.props.id}
          contentType={this.context.comments_contenttype}
        />
        {this.renderDeleteModal()}
        <h3 className={this.props.is_deleted ? 'comment-deleted-author' : 'comment-author'}>
          {this.props.is_deleted ? this.props.user_name
            : <a href={`/profile/${this.props.user_name}`} data-embed-target="external">{this.props.user_name}</a>}
          {this.renderModeratorLabel()}
        </h3>
        {this.renderLastDate()}
        {this.renderComment()}
        <div className="action-bar">
          <nav className="navbar navbar-default navbar-static">
            {this.renderRatingBox()}
            {this.allowForm() &&
              <button
                disabled={this.state.showChildComments}
                className="btn comment-answer-button"
                type="button"
                onClick={this.replyComments.bind(this)}
              >
                <i className="fa fa-reply" aria-hidden="true" /> {answerTag}
              </button>}
            {this.context.isAuthenticated && !this.props.is_deleted &&
              <CommentManageDropdown
                id={this.props.id}
                toggleEdit={this.toggleEdit.bind(this)}
                renderModeratorOptions={(this.isOwner() || this.context.isModerator) && !this.props.isReadOnly}
              />}
          </nav>
        </div>

        {this.props.child_comments && this.props.child_comments.length > 0 &&
          <div className="action-bar">
            <div className="navbar">
              <button className="comment-reply-button" type="button" onClick={this.toggleShowComments.bind(this)}>
                <i className={this.state.showChildComments ? 'fa fa-minus' : 'fa fa-plus'} aria-hidden="true" />
                {getViewRepliesText(this.props.child_comments.length, this.state.showChildComments)}
              </button>
            </div>
          </div>}

        {this.state.showChildComments
          ? <div className="comment-child-list">
            <CommentList
              comments={this.props.child_comments}
              parentIndex={this.props.index}
              handleCommentDelete={this.props.handleCommentDelete}
              handleCommentModify={this.props.handleCommentModify}
              isReadOnly={this.props.isReadOnly}
              handleEditErrorClick={this.props.handleEditErrorClick}
            />
            <CommentForm
              subjectType={this.context.comments_contenttype}
              subjectId={this.props.id}
              onCommentSubmit={this.props.handleCommentSubmit}
              parentIndex={this.props.index}
              placeholder={answerPlaceholderText}
              error={this.props.replyError}
              errorMessage={this.props.errorMessage}
              handleErrorClick={() => this.props.handleReplyErrorClick(this.props.index, this.props.parentIndex)}
              rows="3"
              grabFocus={this.state.replyFormHasFocus}
            />
          </div> : null}
      </div>
    )
  }
}

Comment.contextTypes = {
  comments_contenttype: PropTypes.number,
  isAuthenticated: PropTypes.bool,
  isModerator: PropTypes.bool,
  user_name: PropTypes.string,
  contentType: PropTypes.number
}

module.exports = Comment
