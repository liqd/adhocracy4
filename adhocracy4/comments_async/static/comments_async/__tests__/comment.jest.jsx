import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import Comment from '../comment'

describe('Comment Component', () => {
  const defaultProps = {
    user_name: 'Participant 1',
    created: 'May 21, 2024, 11:44 a.m.',
    modified: null,
    children: 'This is a test comment.',
    id: 1,
    index: 0,
    positiveRatings: 10,
    negativeRatings: 2,
    user_info: { is_users_own_comment: false },
    userRatng: null,
    userRatingId: null,
    child_comments: [],
    onCommentDelete: jest.fn(),
    onCommentSubmit: jest.fn(),
    onCommentModify: jest.fn(),
    onReplyErrorClick: jest.fn(),
    onEditErrorClick: jest.fn(),
    anchoredCommentId: null,
    anchoredCommentParentId: null,
    has_comment_commenting_permission: true,
    hasCommentingPermission: true,
    wouldHaveCommentingPermission: true,
    projectIsPublic: true,
    content_type: 107,
    comment_content_type: 108,
    useTermsOfUse: false,
    agreedTermsOfUse: false,
    orgTermsUrl: '',
    setCommentError: jest.fn(),
    setCommentEditError: jest.fn()
  }
  const childComment = {
    ai_report: null,
    authorIsModerator: false,
    child_comments: [],
    comment: 'This is a test child comment.',
    comment_categories: null,
    comment_content_type: 108,
    content_type: 107,
    created: '2024-08-07T12:34:56Z',
    displayNotification: false,
    editError: null,
    errorMessage: null,
    moderator_feedback: null,
    modified: undefined,
    object_pk: 1,
    has_comment_commenting_permission: true,
    id: 2,
    index: 0,
    is_deleted: false,
    is_removed: false,
    is_cenesored: false,
    is_blocked: false,
    is_moderator_marked: false,
    ratings: {
      positiveRatings: 0,
      negativeRatings: 0,
      userRating: null,
      userRatingId: null
    },
    replyError: null,
    user_info: {
      is_users_own_comment: false,
      authenticated_user_pk: 0,
      has_viewing_permission: true,
      has_rating_permission: true,
      has_changing_permission: false,
      has_deleting_permission: false,
      has_moderating_permission: false,
      has_comment_commenting_permission: true
    },
    user_name: 'Participant 2',
    user_pk: 1,
    user_profile_url: 'https://liqd.net/profile_url',
    user_image: null,
    user_image_fallback: 'https://liqd.net/profile_image'
  }

  test('renders comment with creator and comment text', () => {
    const tree = render(<Comment {...defaultProps} />)
    expect(tree).toMatchSnapshot()

    expect(screen.getByText(/Participant 1/)).toBeInTheDocument()
    expect(screen.getByText(/This is a test comment./)).toBeInTheDocument()
  })

  test('renders comment with share button', () => {
    render(<Comment {...defaultProps} />)

    expect(screen.getByText(/Share$/)).toBeInTheDocument()
  })

  test('renders comment with creation date', () => {
    render(<Comment {...defaultProps} />)

    expect(screen.getByText(/May 21, 2024, 11:44 a.m./)).toBeInTheDocument()
  })

  test('renders comment with modified date', () => {
    render(<Comment {...defaultProps} modified="May 22, 2024, 11:44 a.m./" />)

    expect(
      screen.queryByText(/May 21, 2024, 11:44 a.m./)
    ).not.toBeInTheDocument()
    expect(screen.getByText(/Latest edit on May 22, 2024, 11:44 a.m./)).toBeInTheDocument()
  })

  test('renders comment with moderator badge', () => {
    const tree = render(<Comment {...defaultProps} authorIsModerator />)
    expect(tree).toMatchSnapshot()

    expect(screen.getByText(/Moderator/)).toBeInTheDocument()
  })

  test('renders deleted comment without child comments correctly', () => {
    render(<Comment {...defaultProps} is_deleted />)

    const creator = screen.getByText(/Participant 1/)
    expect(creator.closest('div')).toHaveClass('a4-comments__deleted-author')
    expect(screen.queryByText(/Share$/)).not.toBeInTheDocument()
    expect(screen.queryByText(/Reply/)).not.toBeInTheDocument()
    expect(screen.queryByText(/reply./)).not.toBeInTheDocument()
  })

  test('renders comment without child comments', () => {
    render(<Comment {...defaultProps} />)

    expect(screen.getByText(/Reply/)).toBeInTheDocument()
    expect(screen.queryByText(/reply./)).not.toBeInTheDocument()
  })

  test('renders comment with child comment', () => {
    render(<Comment {...defaultProps} child_comments={[childComment]} />)

    expect(screen.getByText(/1 reply/)).toBeInTheDocument()
    expect(screen.queryByText(/Reply./)).not.toBeInTheDocument()
  })

  test('toggling reply button toggles child comment', () => {
    render(<Comment {...defaultProps} child_comments={[childComment]} />)

    expect(screen.getByText(/1 reply/)).toBeInTheDocument()
    expect(
      screen.queryByText(/This is a test child comment./)
    ).not.toBeInTheDocument()

    const replyButton = screen.getByRole('button', {
      name: '%s replies1 reply'
    })
    fireEvent.click(replyButton)
    expect(
      screen.getByText(/This is a test child comment./)
    ).toBeInTheDocument()
    fireEvent.click(replyButton)
    expect(
      screen.queryByText(/This is a test child comment./)
    ).not.toBeInTheDocument()
  })

  test('displays ratings correctly', () => {
    render(<Comment {...defaultProps} />)

    expect(screen.getByText('10')).toBeInTheDocument()
    expect(screen.getByText('2')).toBeInTheDocument()
  })

  test('displays user rating correctly', () => {
    render(<Comment {...defaultProps} userRating={1} userRatingsId={0} />)

    const positiveRatings = screen.getByText('10')
    expect(positiveRatings).toBeInTheDocument()
    expect(positiveRatings).toHaveClass('is-selected')
    const negativeRatings = screen.getByText('2')
    expect(negativeRatings).toBeInTheDocument()
    expect(negativeRatings).not.toHaveClass('is-selected')
  })
})
