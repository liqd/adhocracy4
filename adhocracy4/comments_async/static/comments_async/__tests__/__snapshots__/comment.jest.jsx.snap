// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Component renders comment with creator and comment text 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li>
        <div
          class="a4-comments__comment"
        >
          <a
            class="a4-comments__anchor"
            href="./?comment=1"
            id="comment_1"
          >
            Comment 1
          </a>
          <div
            class="a4-comments__box"
          >
            <div
              class="row"
            >
              <div
                class="col-2 col-lg-1 a4-comments__user-img"
              />
              <div
                class="col-7 col-md-8 a4-comments__author-container"
              >
                <div
                  class="a4-comments__author"
                >
                  <a>
                    Participant 1
                  </a>
                </div>
                <time
                  class="a4-comments__submission-date"
                >
                  May 21, 2024, 11:44 a.m.
                </time>
              </div>
              <div
                class="col-1 ms-auto a4-comments__dropdown-container"
              />
              <div
                class="col-12"
              >
                <div
                  class="a4-comments__text"
                >
                  <p>
                    This is a test comment.
                  </p>
                </div>
              </div>
              <div
                class="col-6 a4-comments__read-btn-container"
              />
            </div>
          </div>
          <div
            class="a4-comments__action"
          >
            <div
              class="row"
            >
              <div
                class="col-12 a4-comments__action-bar-container"
              >
                <div
                  class="rating"
                  data-testid="rating-box"
                >
                  <button
                    aria-pressed="false"
                    class="rating-button rating-up"
                    disabled=""
                    type="button"
                  >
                    <i
                      aria-hidden="true"
                      class="fa fa-thumbs-up"
                    />
                     
                    10
                     
                    <span
                      class="rating__label"
                    >
                      Likes
                    </span>
                    <span
                      class="a4-sr-only"
                    >
                      Click to like
                    </span>
                  </button>
                  <button
                    aria-pressed="false"
                    class="rating-button rating-down"
                    disabled=""
                    type="button"
                  >
                    <i
                      aria-hidden="true"
                      class="fa fa-thumbs-down"
                    />
                     
                    2
                     
                    <span
                      class="rating__label"
                    >
                      Dislikes
                    </span>
                    <span
                      class="a4-sr-only"
                    >
                      Click to dislike
                    </span>
                  </button>
                </div>
                <div
                  class="a4-comments__action-bar"
                >
                  <button
                    class="btn btn--no-border a4-comments__action-bar__btn"
                    type="button"
                  >
                    <a
                      href="#child-comment-form"
                    >
                      <i
                        aria-hidden="true"
                        class="far fa-comment"
                      />
                       
                      Reply
                    </a>
                  </button>
                  <button
                    class="a4-modal__toggle btn-link link"
                  >
                    <i
                      aria-hidden="true"
                      class="fas fa-share"
                    />
                     
                    Share
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </li>
    </div>
    <dialog
      class="a4-modal"
      id=":r0:"
    >
      <div
        class="a4-modal__content"
      >
        <header
          class="a4-modal__header"
        >
          <h2
            class="a4-modal__title"
          >
            Share link
          </h2>
          <button
            aria-label="Close"
            class="a4-modal__close"
            type="button"
          >
            <i
              aria-hidden="true"
              class="fa fa-times"
            />
          </button>
        </header>
        <div
          class="a4-modal__body "
        >
          <div
            class="input-group a4-url-modal__body"
          >
            <input
              class="form-control"
              id="share-url-undefined"
              readonly=""
              type="text"
              value="http://localhost/?comment=1"
            />
            <button
              aria-pressed="false"
              autocomplete="off"
              class="a4-url-modal__button"
            >
              Copy
            </button>
          </div>
        </div>
      </div>
    </dialog>
  </body>,
  "container": <div>
    <li>
      <div
        class="a4-comments__comment"
      >
        <a
          class="a4-comments__anchor"
          href="./?comment=1"
          id="comment_1"
        >
          Comment 1
        </a>
        <div
          class="a4-comments__box"
        >
          <div
            class="row"
          >
            <div
              class="col-2 col-lg-1 a4-comments__user-img"
            />
            <div
              class="col-7 col-md-8 a4-comments__author-container"
            >
              <div
                class="a4-comments__author"
              >
                <a>
                  Participant 1
                </a>
              </div>
              <time
                class="a4-comments__submission-date"
              >
                May 21, 2024, 11:44 a.m.
              </time>
            </div>
            <div
              class="col-1 ms-auto a4-comments__dropdown-container"
            />
            <div
              class="col-12"
            >
              <div
                class="a4-comments__text"
              >
                <p>
                  This is a test comment.
                </p>
              </div>
            </div>
            <div
              class="col-6 a4-comments__read-btn-container"
            />
          </div>
        </div>
        <div
          class="a4-comments__action"
        >
          <div
            class="row"
          >
            <div
              class="col-12 a4-comments__action-bar-container"
            >
              <div
                class="rating"
                data-testid="rating-box"
              >
                <button
                  aria-pressed="false"
                  class="rating-button rating-up"
                  disabled=""
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="fa fa-thumbs-up"
                  />
                   
                  10
                   
                  <span
                    class="rating__label"
                  >
                    Likes
                  </span>
                  <span
                    class="a4-sr-only"
                  >
                    Click to like
                  </span>
                </button>
                <button
                  aria-pressed="false"
                  class="rating-button rating-down"
                  disabled=""
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="fa fa-thumbs-down"
                  />
                   
                  2
                   
                  <span
                    class="rating__label"
                  >
                    Dislikes
                  </span>
                  <span
                    class="a4-sr-only"
                  >
                    Click to dislike
                  </span>
                </button>
              </div>
              <div
                class="a4-comments__action-bar"
              >
                <button
                  class="btn btn--no-border a4-comments__action-bar__btn"
                  type="button"
                >
                  <a
                    href="#child-comment-form"
                  >
                    <i
                      aria-hidden="true"
                      class="far fa-comment"
                    />
                     
                    Reply
                  </a>
                </button>
                <button
                  class="a4-modal__toggle btn-link link"
                >
                  <i
                    aria-hidden="true"
                    class="fas fa-share"
                  />
                   
                  Share
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Comment Component renders comment with moderator badge 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li>
        <div
          class="a4-comments__comment"
        >
          <a
            class="a4-comments__anchor"
            href="./?comment=1"
            id="comment_1"
          >
            Comment 1
          </a>
          <div
            class="a4-comments__box"
          >
            <div
              class="row"
            >
              <div
                class="col-2 col-lg-1 a4-comments__user-img"
              />
              <div
                class="col-7 col-md-8 a4-comments__author-container"
              >
                <div
                  class="a4-comments__author"
                >
                  <a>
                    Participant 1
                  </a>
                </div>
                <span
                  class="a4-comments__moderator"
                >
                  Moderator
                </span>
                <time
                  class="a4-comments__submission-date"
                >
                  May 21, 2024, 11:44 a.m.
                </time>
              </div>
              <div
                class="col-1 ms-auto a4-comments__dropdown-container"
              />
              <div
                class="col-12"
              >
                <div
                  class="a4-comments__text"
                >
                  <p>
                    This is a test comment.
                  </p>
                </div>
              </div>
              <div
                class="col-6 a4-comments__read-btn-container"
              />
            </div>
          </div>
          <div
            class="a4-comments__action"
          >
            <div
              class="row"
            >
              <div
                class="col-12 a4-comments__action-bar-container"
              >
                <div
                  class="rating"
                  data-testid="rating-box"
                >
                  <button
                    aria-pressed="false"
                    class="rating-button rating-up"
                    disabled=""
                    type="button"
                  >
                    <i
                      aria-hidden="true"
                      class="fa fa-thumbs-up"
                    />
                     
                    10
                     
                    <span
                      class="rating__label"
                    >
                      Likes
                    </span>
                    <span
                      class="a4-sr-only"
                    >
                      Click to like
                    </span>
                  </button>
                  <button
                    aria-pressed="false"
                    class="rating-button rating-down"
                    disabled=""
                    type="button"
                  >
                    <i
                      aria-hidden="true"
                      class="fa fa-thumbs-down"
                    />
                     
                    2
                     
                    <span
                      class="rating__label"
                    >
                      Dislikes
                    </span>
                    <span
                      class="a4-sr-only"
                    >
                      Click to dislike
                    </span>
                  </button>
                </div>
                <div
                  class="a4-comments__action-bar"
                >
                  <button
                    class="btn btn--no-border a4-comments__action-bar__btn"
                    type="button"
                  >
                    <a
                      href="#child-comment-form"
                    >
                      <i
                        aria-hidden="true"
                        class="far fa-comment"
                      />
                       
                      Reply
                    </a>
                  </button>
                  <button
                    class="a4-modal__toggle btn-link link"
                  >
                    <i
                      aria-hidden="true"
                      class="fas fa-share"
                    />
                     
                    Share
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </li>
    </div>
    <dialog
      class="a4-modal"
      id=":r4:"
    >
      <div
        class="a4-modal__content"
      >
        <header
          class="a4-modal__header"
        >
          <h2
            class="a4-modal__title"
          >
            Share link
          </h2>
          <button
            aria-label="Close"
            class="a4-modal__close"
            type="button"
          >
            <i
              aria-hidden="true"
              class="fa fa-times"
            />
          </button>
        </header>
        <div
          class="a4-modal__body "
        >
          <div
            class="input-group a4-url-modal__body"
          >
            <input
              class="form-control"
              id="share-url-undefined"
              readonly=""
              type="text"
              value="http://localhost/?comment=1"
            />
            <button
              aria-pressed="false"
              autocomplete="off"
              class="a4-url-modal__button"
            >
              Copy
            </button>
          </div>
        </div>
      </div>
    </dialog>
  </body>,
  "container": <div>
    <li>
      <div
        class="a4-comments__comment"
      >
        <a
          class="a4-comments__anchor"
          href="./?comment=1"
          id="comment_1"
        >
          Comment 1
        </a>
        <div
          class="a4-comments__box"
        >
          <div
            class="row"
          >
            <div
              class="col-2 col-lg-1 a4-comments__user-img"
            />
            <div
              class="col-7 col-md-8 a4-comments__author-container"
            >
              <div
                class="a4-comments__author"
              >
                <a>
                  Participant 1
                </a>
              </div>
              <span
                class="a4-comments__moderator"
              >
                Moderator
              </span>
              <time
                class="a4-comments__submission-date"
              >
                May 21, 2024, 11:44 a.m.
              </time>
            </div>
            <div
              class="col-1 ms-auto a4-comments__dropdown-container"
            />
            <div
              class="col-12"
            >
              <div
                class="a4-comments__text"
              >
                <p>
                  This is a test comment.
                </p>
              </div>
            </div>
            <div
              class="col-6 a4-comments__read-btn-container"
            />
          </div>
        </div>
        <div
          class="a4-comments__action"
        >
          <div
            class="row"
          >
            <div
              class="col-12 a4-comments__action-bar-container"
            >
              <div
                class="rating"
                data-testid="rating-box"
              >
                <button
                  aria-pressed="false"
                  class="rating-button rating-up"
                  disabled=""
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="fa fa-thumbs-up"
                  />
                   
                  10
                   
                  <span
                    class="rating__label"
                  >
                    Likes
                  </span>
                  <span
                    class="a4-sr-only"
                  >
                    Click to like
                  </span>
                </button>
                <button
                  aria-pressed="false"
                  class="rating-button rating-down"
                  disabled=""
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="fa fa-thumbs-down"
                  />
                   
                  2
                   
                  <span
                    class="rating__label"
                  >
                    Dislikes
                  </span>
                  <span
                    class="a4-sr-only"
                  >
                    Click to dislike
                  </span>
                </button>
              </div>
              <div
                class="a4-comments__action-bar"
              >
                <button
                  class="btn btn--no-border a4-comments__action-bar__btn"
                  type="button"
                >
                  <a
                    href="#child-comment-form"
                  >
                    <i
                      aria-hidden="true"
                      class="far fa-comment"
                    />
                     
                    Reply
                  </a>
                </button>
                <button
                  class="a4-modal__toggle btn-link link"
                >
                  <i
                    aria-hidden="true"
                    class="fas fa-share"
                  />
                   
                  Share
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
